server/keyspace/keyspace.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return nil, err
+	}
+	// TODO: Enable Transaction at storage layer to save MetaData and NameToID in a single transaction.
+	// Create and save keyspace metadata.
+	keyspace := &keyspacepb.KeyspaceMeta{
+		Id:             newID,
+		Name:           request.Name,
+		State:          keyspacepb.KeyspaceState_ENABLED,
+		CreatedAt:      request.Now.Unix(),
+		StateChangedAt: request.Now.Unix(),
+		Config:         request.Config,
+	}
+	manager.metaLock.Lock()
+	defer manager.metaLock.Unlock()
+	// Check if keyspace with that id already exists.
+	keyspaceExists, err := manager.store.LoadKeyspace(newID, &keyspacepb.KeyspaceMeta{})
+	if err != nil {
+		return nil, err
+	}
+	if keyspaceExists {
+		return nil, ErrKeyspaceExists
+	}
+	// Save keyspace meta before saving id.
+	if err = manager.store.SaveKeyspace(keyspace); err != nil {
+		return nil, err
+	}
+	// Create name to ID entry,
+	// if this failed, previously stored keyspace meta should be removed.
+	if err = manager.createNameToID(newID, request.Name); err != nil {
+		if removeErr := manager.store.RemoveKeyspace(newID); removeErr != nil {
+			return nil, errors.Wrap(removeErr, "failed to remove keyspace meta after save spaceID failure")
+		}
+		return nil, err
+	}
+
+	return keyspace, nil
+}
+
+// LoadKeyspace returns the keyspace specified by name.
+// It returns error if loading or unmarshalling met error or if keyspace does not exist.
+func (manager *Manager) LoadKeyspace(name string) (*keyspacepb.KeyspaceMeta, error) {
+	// First get keyspace ID from the name given.
+	loaded, spaceID, err := manager.store.LoadKeyspaceIDByName(name)
+	if err != nil {
+		return nil, err
+	}
+	if !loaded {
+		return nil, ErrKeyspaceNotFound
+	}
+	// Load the keyspace with target ID.
+	keyspace := &keyspacepb.KeyspaceMeta{}
+	loaded, err = manager.store.LoadKeyspace(spaceID, keyspace)
+	if err != nil {
+		return nil, err
+	}
+	if !loaded {
+		return nil, ErrKeyspaceNotFound
+	}
+	return keyspace, nil
+}
+
+// Mutation represents a single operation to be applied on keyspace config.
+type Mutation struct {
+	Op    OpType
+	Key   string
+	Value string
+}
+
+// OpType defines the type of keyspace config operation.
+type OpType int
+
+const (
+	// OpPut denotes a put operation onto the given config.
+	// If target key exists, it will put a new value,
+	// otherwise, it creates a new config entry.
+	OpPut OpType = iota + 1 // Operation type starts at 1.
+	// OpDel denotes a deletion operation onto the given config.
+	// Note: OpDel is idempotent, deleting a non-existing key
+	// will not result in error.
+	OpDel
+)
+
+// UpdateKeyspaceConfig changes target keyspace's config in the order specified in mutations.
+// It returns error if saving failed, operation not allowed, or if keyspace not exists.
+func (manager *Manager) UpdateKeyspaceConfig(name string, mutations []*Mutation) (*keyspacepb.KeyspaceMeta, error) {
+	manager.metaLock.Lock()
+	defer manager.metaLock.Unlock()
```

Comment:
I think it is not a good idea to share a lock for all key-spaces.

Reply:
I changed the global keyspace lock to a map of mutexes (`lockGroup`) that uses the lower byte of spaceID as key.

This should levitate the lock contention, especially during consecutive keyspace creation (since spaceIDs are allocated sequentially), and avoid using too much memory to create a mutex for every keyspace. 

It's implementation and corresponding tests are located in `keyspace/util.go` & `keyspace/util_test.go`, PTAL and let me know what you think.